### ISSUE TITLE
Introduction of Sample Weights

### ISSUE THREAD

**Issue by JohnStott** [manifoldai/merf#18]
Hi, I am looking to extend the functionality to allow a sample_weight vector to be passed in...  sklearn.ensemble.RandomForestRegressor requires the sample_weight as a separate vector so that it can naturally work out the variance impurity (MSE based) and node averages correctly.  

I am wondering whether I could simply use y * sample_weight as y throughout the merf.fit() function and then extend the RF fit to something like `rf.fit(X, y_star / sample_weight, sample_weight)`?  I assume this would work from what I understand of the current implementation, since it appears that the normalisation doesn't consider cluster sizes and thus there should be no need to weight here (unfortunately I don't have access to the MERF paper):

```
# Normalize the sums to get sigma2_hat and D_hat
sigma2_hat = (1.0 / n_obs) * sigma2_hat_sum
D_hat = (1.0 / n_clusters) * D_hat_sum
```

To clarify, I wouldn't expect D_hat to be normalised as above if the cluster size (or in my case, cluster weighting) was important to the overall process?  If it had have been important, I would have expected a D_hat per cluster (i.e., in proportion to how many samples the cluster held versus the total samples).

**Comment by resdntalien** [comment]
@JohnStott Sorry I am getting to this so late. I just merged a PR that exposes all the parameters of the random forest to the user. Also, planning on generalizing this to any estimator soon. Would that fix your problem?

**Comment by JoGall** [comment]
Sorry if I'm missing something obvious here but could you please tell me how exactly I can implement the sample weights? I've tried all three ways I can think of, i.e.:

```
# 1
rf_model = RandomForestRegressor(sample_weight=sample_weights)
merf = MERF(rf_model)
merf.fit(X=X_train, Z=Z_train, clusters=clusters_train, y=y_train)

# 2
rf_model = RandomForestRegressor()
merf = MERF(rf_model, sample_weight=sample_weights)
merf.fit(X=X_train, Z=Z_train, clusters=clusters_train, y=y_train)

# 3
rf_model = RandomForestRegressor()
merf = MERF(rf_model)
merf.fit(X=X_train, Z=Z_train, clusters=clusters_train, y=y_train, sample_weight=sample_weights)
```

I can see the `sample_weight` parameter defined [here](https://github.com/scikit-learn/scikit-learn/blob/24593310d10e13efc364d17ce457dce0c7fa5df0/sklearn/ensemble/_forest.py) though, should I be importing this version somehow instead of `sklearn.ensemble.RandomForestRegressor`?

Thanks for your work!
